RFC-0001-Agile Coretime.md

```diff
@@ -0,0 +1,298 @@
+
+1. The purchase always succeeds and as such MUST be processed prior to regular `purchase` orders.
+2. There MUST be an ongoing legacy parachain lease whose parachain is assigned to `core` and which is due to expire during the Coretime period being purchased.
+
+AllowedRenewals map is updated with this information ready for the following Bulk sale.
+
+### Notes on Price Setting
+
+The specific price setting mechanism is out of scope for this proposal and should be covered in some other work. The present proposal assumes the existence of a price-setting mechanism which could take into account three parameters; two mostly fixed and one which changes each `BULK_PERIOD`. These parameters are `BULK_TARGET`, `BULK_LIMIT` and `CORES_SOLD` which is the actual number of cores sold for the present `BULK_PERIOD` and is always an unsigned integer at most `BULK_LIMIT`.
+
+In general we would expect the price to increase the closer `CORES_SOLD` gets to `BULK_LIMIT` and to decrease the closer it gets to zero. If it is exactly equal to `BULK_TARGET`, then we would expect the price to remain the same.
+
+A simple example of this would be the formula:
+
+```
+NEW_PRICE := IF CORES_OLD < BULK_TARGET THEN
+    OLD_PRICE - OLD_PRICE / 2 * CORES_SOLD / BULK_TARGET
+ELSE
+    OLD_PRICE + OLD_PRICE / 2 *
+        (CORES_SOLD - BULK_TARGET) / (BULK_LIMIT - BULK_TARGET)
+END IF
+```
+
+This exists only as a trivial example to demonstrate a basic solution exists, and should not be intended as a concrete proposal.
+
+### Notes on Instantaneous Core Sales
+
+For access to the Instantaneous Core sales we may include an `allocate_instantaneous` function. This should allocate the Coretime for usage by Polkadot to serve instantaneous requests and allow the `owner` to collect a pro-rata amount of total Instantaneous sales revenue.
+
+For an efficient market to form around the provision of Bulk-purchased Cores into the pool of cores available for Instantaneous Coretime purchase, it is crucial to ensure that price changes for the purchase of Instantaneous Coretime are reflected well in the revenues of private Coretime providers during the same period.
+
+In order to ensure this, then it is crucial that Instantaneous Coretime, once purchased, cannot be held indefinitely prior to eventual use since, if this were the case, a nefarious collator could purchase Coretime when cheap and utilize it some time later when expensive and deprive private Coretime providers of their revenue. It SHOULD therefore be assumed that Instantaneous Coretime, once purchased, has a definite and short "shelf-life", after which it becomes unusable. This incentivizes collators to avoid purchasing Coretime unless they expect to utilize it imminently and thus helps create an efficient market-feedback mechanism whereby a higher price will actually result in material revenues for private Coretime providers who contribute to the pool of Cores available to service Instantaneous Coretime purchases.
```

Comment:
in the current implementation, purchased is equivalent to queued. Hence it will be served as soon as possible.

Reply:
Note that, as I understand it, the broker chain doesn't attempt to do anything related to on-demand / Instantaneous except tell the relay-chain that certain cores (or fractions of cores) are to be allocated into the on-demand mechanism for some time period, and expects some information about how much revenue was accrued during that time period on the core to be communicated back to the broker chain or readable from it.

We're still discussing the exact interface that the broker chain would use to communicate with the relay chain, but it'd be something like

`RelayChain::allocate(core_index, start, Option<end>, allocation)` where the `allocation: Vec<ParaId>` can encode information about allocating some or all of the coretime to the on-demand pool.

Reply:
(See https://github.com/polkadot-fellows/RFCs/pull/5)